Add requirement for GetURL onto ServiceClients (#5996)
Azure/azure-sdk: docs/cpp/introduction.md

{% include requirement/MUST id="cpp-service-client-name" %} name service client types with the _Client_ suffix (for example, `ConfigurationClient`).

{% include requirement/MUST id="cpp-service-client-namespace" %} place service client types that the consumer is most likely to interact with in the root namespace of the client library (for example, `Azure::<group>::<service>`). Specialized service clients should be placed in sub-packages.

{% include requirement/MUST id="cpp-service-client-type" %} make service clients `classes`, not `structs`.

{% include requirement/MUST id="cpp-service-client-immutable" %} ensure that all service client classes thread safe (usually by making them immutable and stateless).

{% include requirement/MUST id="cpp-service-client-geturl" %} expose a `GetUrl()` method which returns the URL.

#### Service Client Constructors {#cpp-client-ctor}

{% include requirement/MUST id="cpp-service-client-constructor-minimal" %} provide a minimal constructor that takes only the parameters required to connect to the service.

> For example, you may use a connection string, or host name and authentication.  It should be easy to start using the client without extensive customization.

Reviewer: ```
It should be easy to start using the client without extensive customization.
```
I'm not 100% sure what this is trying to say, or what the relationship between this and the connection string is.

Also: Do we support clients which support host name's directly (as opposed to URLs?)

### Service Client {#cpp-client}

Service clients are the main starting points for developers calling Azure services with the Azure SDK. Each client library should have at least one client in its main namespace, so it’s easy to discover. The guidelines in this section describe patterns for the design of a service client.

There exists a distinction that must be made clear with service clients: not all classes that perform HTTP (or otherwise) requests to a service are automatically designated as a service client. A service client designation is only applied to classes that are able to be directly constructed because they are uniquely represented on the service. Additionally, a service client designation is only applied if there is a specific scenario that applies where the direct creation of the client is appropriate. If a resource can not be uniquely identified or there is no need for direct creation of the type, then the service client designation should not apply.

{% include requirement/MUST id="cpp-service-client-name" %} name service client types with the _Client_ suffix (for example, `ConfigurationClient`).

> {% include requirement/MUST id="cpp-service-client-namespace" %} place service client types that the consumer is most likely to interact with in the root namespace of the client library (for example, `Azure::<group>::KeyVault`). Specialized service clients should be placed in sub-packages.

Reviewer 2: I don't think it is true that we want service client types at the root namespace like `Azure::KeyVault`. For example, `KeyClient` lives in `Azure::KeyVault::Keys`.

The snippet in .NET and Java highlight that well. We could borrow some of the verbiage from there:
`Azure.<group>.<service_name>`

https://azure.github.io/azure-sdk/java_introduction.html#service-client
https://azure.github.io/azure-sdk/dotnet_introduction.html#dotnet-client

Reviewer 2: An example code snippet that follows the DOs/DO NOTs would make things much clearer.

Author: KeyVault was a bad choice for my example.  Adjusted

{% include requirement/MUST id="cpp-service-client-type" %} make service clients `classes`, not `structs`.

> {% include requirement/MUST id="cpp-service-client-immutable" %} ensure that all service client classes are immutable and stateless upon instantiation.

Reviewer 3: ```suggestion
{% include requirement/MUST id="cpp-service-client-immutable" %} ensure that all service client classes are thread safe (usually by making them immutable and stateless) .
```

Reviewer 2: Removed space before the period in your suggestion.

> {% include requirement/MUST id="cpp-service-client-geturl" %} expose a `GetURL()` method which returns the URL.

Reviewer 2: Fix casing
```suggestion
{% include requirement/MUST id="cpp-service-client-geturl" %} expose a `GetUrl()` method which returns the URL.
```

> {% include requirement/MUST id="cpp-service-client-geturl" %} expose a `GetURL()` method which returns the URL.

Reviewer 2: Fix casing
```suggestion
{% include requirement/MUST id="cpp-service-client-geturl" %} expose a `GetUrl()` method which returns the URL.
```

#### Service Client Constructors {#cpp-client-ctor}

> {% include requirement/MUST id="cpp-service-client-constructor-minimal" %} provide a minimal constructor that takes only the parameters required to connect to the service.

Reviewer 3: Usually a ctor takes an endpoint URL, a credential type, and optional options.

Reviewer 3: It's true that a very few Clients can be created with an anonymous credential and the ctor does not need the credential parameter in this case. It would still need the endpoint and ClientOptions.

Reviewer: FWIW, Attestation has both authenticated and anonymous methods, so an AttestationClient can be created with or without a TokenCredential object.

> For example, you may use a connection string, or host name and authentication.  It should be easy to start using the client without extensive customization.

Reviewer 3: Usually methods that construct a client from a connection string are static factory methods; not constructors.

Author: Do we want to mandate that?  I can take away the for example portion of this.

Reviewer 3: Yes, I think it best to mandate this for a few reasons: We want to discourage use of connection strings and static factory methods are a little harder to discover than ctors (which is a good thing). Also, if a ctor takes a connection string, creds, and options then it conflicts with a ctor that takes an endpoint string, creds and options - we must be able to tell them apart from one another.

Author: Ok, I'll add that in a follow up  PR.

> {% include requirement/MUSTNOT id="cpp-client-constructor-no-default-params" %} use default parameters in the simplest constructor.

Reviewer 3: This would just be the optional options

> {% include requirement/MUST id="cpp-client-constructor-overloads" %} provide constructor overloads that allow specifying additional options such as credentials, a custom HTTP pipeline, or advanced configuration.

Reviewer 3: We usually do not do this. Creds are an explicit parameter and pipeline customizations are accomplished via ClientOption fields

Reviewer 2: Beyond the fact that we **don't** do this, our HTTP pipeline isn't public atm, so we **can't** do it either.

> Custom pipeline and client-specific configuration are represented by an `options` parameter. The type of the parameter is typically a subclass of ```ClientOptions``` type, shown below.

Reviewer 3: Yes, this is correct. Not what is above.

Reviewer 2: This might be confusing given HttpPipeline is internal only. We don't have custom pipeline as an options parameter.

Author: Fixed

{% include requirement/MUSTNOT id="cpp-client-constructor-no-default-params" %} use default parameters in the simplest constructor.

> {% include requirement/MUST id="cpp-client-constructor-overloads" %} provide constructor overloads that allow specifying additional options via  an `options` parameter. The type of the parameter is typically a subclass of ```ClientOptions``` type, shown below.

Reviewer 2: nit: extra space between words
```suggestion
{% include requirement/MUST id="cpp-client-constructor-overloads" %} provide constructor overloads that allow specifying additional options via an `options` parameter. The type of the parameter is typically a subclass of ```ClientOptions``` type, shown below.
```
